### PR TITLE
Add shard compatibility to `merge_str_prep.py`

### PR DESCRIPTION
**Pull request by hopedisastro**

### CODE REVIEW

**Comment by hopedisastro** [comment]
Updates merge_str_prep.py so that it is compatible with sharded VCFs.

**Comment by hopedisastro** [comment]
Remaining lint conflicts relate to other scripts in the repo, so have been ignored.

**Review by MattWellie — APPROVED** [review]

File: str/trtools/merge_str_prep.py [hunk 1]
```diff
@@ -22,20 +23,21 @@
...
+@click.option('--sharded', is_flag=True, help='Assume a sharded catalog was used')
 # input sample ID
 @click.argument('internal-wgs-ids', nargs=-1)
 @click.command()
 def main(
-    caller, input_dir, internal_wgs_ids: list[str]
+    caller, input_dir, internal_wgs_ids: list[str], sharded
 ):  # pylint: disable=missing-function-docstring
     # Initializing Batch
     b = get_batch()
```
Comment (illusional) [inline str/trtools/merge_str_prep.py:43]: Can you give your workflow a name - save the blank Hail Batch rows

File: str/trtools/merge_str_prep.py [hunk 2]
```diff
@@ -22,20 +23,21 @@
...
 # input directory
 @click.option('--input-dir', help='gs://...')
+# sharded flag
+@click.option('--sharded', is_flag=True, help='Assume a sharded catalog was used')
 # input sample ID
 @click.argument('internal-wgs-ids', nargs=-1)
 @click.command()
 def main(
-    caller, input_dir, internal_wgs_ids: list[str]
+    caller, input_dir, internal_wgs_ids: list[str], sharded
```
Comment (illusional) [inline str/trtools/merge_str_prep.py:40]:
```suggestion
    caller, input_dir, internal_wgs_ids: list[str], sharded: bool
```

File: str/trtools/merge_str_prep.py [hunk 3]
```diff
@@ -22,20 +23,21 @@
 BCFTOOLS_IMAGE = config['images']['bcftools']
```
Comment (illusional) [inline str/trtools/merge_str_prep.py:23]: I think you can use `images_path('bcftools')` from cpg_utils here?

File: str/trtools/merge_str_prep.py [hunk 4]
```diff
@@ -52,57 +54,66 @@ def main(
...
+                bcftools_job.declare_resource_group(
+                    vcf_sorted={
+                        'reheader.vcf.gz': '{root}.reheader.vcf.gz',
+                        'reheader.vcf.gz.tbi': '{root}.reheader.vcf.gz.tbi',
+                    }
+                )
+                bcftools_job.command(
+                    f"""
+
+                    bcftools reheader -f {ref.fai} {vcf_input} | bcftools sort --temp-dir $BATCH_TMPDIR/ | bgzip -c >  {bcftools_job.vcf_sorted['reheader.vcf.gz']}
```
Comment (illusional) [inline str/trtools/merge_str_prep.py:87]: I'm in favour of breaking these into different lines as it helps readability (in Python and Hail Batch)

File: str/trtools/merge_str_prep.py [hunk 5]
```diff
@@ -52,57 +54,66 @@ def main(
...
-            b.write_output(bcftools_job.vcf_sorted, output_path_gangstr)
+        for vcf_file in input_vcf_dict[id]:
+            vcf_input = b.read_input(vcf_file)
+            bcftools_job = b.new_job(name=f'{vcf_file} Files prep')
+            bcftools_job.image(BCFTOOLS_IMAGE)
+            bcftools_job.cpu(4)
+            bcftools_job.storage('15G')
+
+            if caller == 'eh':
+                bcftools_job.declare_resource_group(
```
Comment (illusional) [inline str/trtools/merge_str_prep.py:78]: Is it worth moving these into their own functions, one for each caller, makes it more readable, especially with this new nesting.

File: str/trtools/merge_str_prep.py [hunk 6]
```diff
@@ -4,7 +4,7 @@
 This script prepares GangSTR/EH VCF files for input into mergeSTR.
 Required input: --caller, --input-dir, and external sample IDs
 For example:
-analysis-runner --access-level test --dataset tob-wgs --description 'tester' --output-dir 'str/5M_run_combined_vcfs/merge_str_prep' merge_str_prep.py --caller=eh --input-dir=gs://cpg-tob-wgs-test/str/5M_run_combined_vcfs CPGtestersorted2
+analysis-runner --access-level test --dataset tob-wgs --description 'tester' --output-dir 'str/5M_run_combined_vcfs/merge_str_prep/v2' merge_str_prep.py --caller=eh --input-dir=gs://cpg-tob-wgs-test/hoptan-str --sharded sharded_tester
```
Comment (illusional) [inline str/trtools/merge_str_prep.py:7]: Could make this more readable:

```suggestion
analysis-runner \
    --access-level test \
    --dataset tob-wgs \
    --description 'tester' \
    --output-dir 'str/5M_run_combined_vcfs/merge_str_prep/v2' \
    merge_str_prep.py \
        --caller=eh \
        --input-dir=gs://cpg-tob-wgs-test/hoptan-str \
        --sharded sharded_tester
```

**Review by illusional — COMMENTED** [review]
Already merged, but some passing comments

**Pull request merged**